File: solutions/php/spiral-matrix/1/SpiralMatrix.php
<?php
// For my Shree DR.MDD

declare(strict_types=1);

class SpiralMatrix
{
    public function draw(int $size): array
    {
        $grid = [];
        for ($rowIdx = 0; $rowIdx < $size; $rowIdx++) {
            $rowArr = [];
            for ($colIdx = 0; $colIdx < $size; $colIdx++) {
                $rowArr[] = 0;
            }
            $grid[] = $rowArr;
        }

        [$currX, $currY, $dirX, $dirY] = [0, 0, 1, 0];

        for ($num = 1; $num <= $size * $size; $num++) {
            $grid[$currY][$currX] = $num;

            if ($currY + $dirY >= $size || $currX + $dirX >= $size || $grid[$currY + $dirY][$currX + $dirX] !== 0) {
                if ($dirX === 0) {
                    $dirX = -$dirY;
                    $dirY = 0;
                } else {
                    $dirY = $dirX;
                    $dirX = 0;
                }
            }

            $currX += $dirX;
            $currY += $dirY;
        }

        return $grid;
    }
}

File: solutions/go/spiral-matrix/1/spiral_matrix.go
// For my Shree DR.MDD
package spiralmatrix

func SpiralMatrix(n int) [][]int {
	grid := make([][]int, n)
	for i := range grid {
		grid[i] = make([]int, n)
	}

	r, c := 0, 0
	dr, dc := 0, 1

	for val := 1; val <= n*n; val++ {
		grid[r][c] = val
		nr, nc := r-dr, c+dc
		if outOfRange(n, nr, nc) || grid[nr][nc] != 0 {
			dc, dr = dr, -dc
		}
		c += dc
		r -= dr
	}
	return grid
}

func outOfRange(size int, coords ...int) bool {
	for _, x := range coords {
		if x < 0 || x >= size {
			return true
		}
	}
	return false
}
